sympy/physics/mechanics/tests/test_joint.py

```diff
@@ -176,7 +177,7 @@ def test_pinjoint_arbitrary_axis():
     assert A.ang_vel_in(N).express(A) == omega * A.y
     assert A.ang_vel_in(N).magnitude() == sqrt(omega**2)
     angle = A.ang_vel_in(N).angle_between(A.y)
-    assert expand_mul(angle).xreplace({omega: 1}) == 0
+    assert angle.xreplace({omega: 1}) == 0
```

Comment:
I'm not sure why this needed to be changed. The `expand_mul` is not needed and doesn't seem to do anything to the expression with or without symengine. However with symengine after `expand_mul` the call to `xreplace` fails:
```python
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /Users/enojb/current/sympy/sympy/sympy/physics/mechanics/tests/test_joint.py(181)test_pinjoint_arbitrary_axis()
-> assert angle.xreplace({omega: 1}) == 0
(Pdb) p angle
acos(omega_J(t)/sqrt(omega_J(t)**2))
(Pdb) p expand_mul(angle)
acos(omega_J(t)/sqrt(omega_J(t)**2))
(Pdb) p angle.xreplace({omega:1})
0
(Pdb) p expand_mul(angle).xreplace({omega:1})
acos(omega_J(t)/sqrt(omega_J(t)**2))
```

Reply:
Oh, it's because `expand_mul` sympifies the expression from a SymEngine expression to a SymPy expression:
```python
(Pdb) p type(angle)
<class 'symengine.lib.symengine_wrapper.acos'>
(Pdb) p type(expand_mul(angle))
acos
```
If we also apply `expand_mul` to `omega` then xreplace works:
```python
(Pdb) p expand_mul(angle).xreplace({expand_mul(omega):1})
0
(Pdb) p expand_mul(angle).xreplace({omega:1})
acos(omega_J(t)/sqrt(omega_J(t)**2))
```
Looks like `xreplace` does not sympify its args:
https://github.com/sympy/sympy/blob/281b1904acc9e15ed04402cff33ebd6ef7812026/sympy/core/basic.py#L1081-L1166

Reply:
A SymEngine expression compares equal to a SymPy expression via `sympify`. However the hashes don't match so they are not identified as equivalent by dict/set:
```python
(Pdb) p angle == expand_mul(angle)
True
(Pdb) p angle in {expand_mul(angle)}
False
(Pdb) hash(angle)
879987935
(Pdb) hash(expand_mul(angle))
-1535576704778178373
```
This breaks the hash invariant:
```
x = y    -->    hash(x) = hash(y)
```